src/tss2-policy/tss2_policy.c

```diff
@@ -0,0 +1,464 @@
+
+    policy_check_not_null(policy_ctx);
+
+    if (exec_callbacks) {
+        policy_ctx->exec_callbacks = *exec_callbacks;
+    } else {
+        memset(&policy_ctx->exec_callbacks, 0, sizeof(policy_ctx->exec_callbacks));
+    }
+
+    return TSS2_RC_SUCCESS;
+}
+
+/** Execute a policy on a given policy session.
+ *
+ * @param[in] policy_ctx The policy context from Tss2_PolicyInstantiate.
+ * @param[in] esys_ctx, the ESAPI context to use for policy execution.
+ * @param[in] hash_alg The hash algorithm used for the policy computation.
+ * @param[in] The ESYS_TR for the policy session to execute policy commands on.
+ *
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_FAPI_RC_AUTHORIZATION_UNKNOWN If the callback for branch selection is
+ *         not defined. This callback will be needed if or policies have to be
+ *         executed.
+ * @retval TSS2_FAPI_RC_BAD_VALUE If the computed branch index deliverd by the
+ *         callback does not identify a branch or If wrong values are detected during execution.
+ * @retval TSS2_FAPI_RC_MEMORY if not enough memory can be allocated.
+ * @retval TSS2_FAPI_RC_BAD_REFERENCE a invalid null pointer is passed.
+ * @retval TSS2_FAPI_RC_IO_ERROR if an error occurs reading the policy file.
+ * @retval TSS2_FAPI_RC_BAD_TEMPLATE In a invalid policy is loaded during execution.
+ * @retval TSS2_FAPI_RC_GENERAL_FAILURE if an internal error occurred.
+ * @retval TSS2_FAPI_RC_AUTHORIZATION_UNKNOWN if a required authorization callback
+ *         is not set.
+ * @retval TSS2_ESYS_RC_* possible error codes of ESAPI.
+ */
+TSS2_RC
+Tss2_PolicyExecute(
+    TSS2_POLICY_CTX *policy_ctx,
+    ESYS_CONTEXT *esys_ctx,
+    ESYS_TR session)
+{
+    policy_check_not_null(policy_ctx);
+    policy_check_not_null(esys_ctx);
+
+    LOG_TRACE("called for policy_path(%s)",
+            policy_ctx->path);
+
+    TSS2_RC r = TSS2_POLICY_RC_GENERAL_FAILURE;
+
+    if (!policy_ctx->is_calculated) {
+        r = Tss2_PolicyCalculate(policy_ctx);
+        return_if_error(r, "Could not calculate policy");
+    }
+
+    enum IFAPI_STATE_POLICY state = POLICY_INIT;
+    IFAPI_POLICY_EXEC_CTX context = { 0 };
+    IFAPI_POLICY_EVAL_INST_CTX eval_ctx = { 0 };
+    IFAPI_IO io = { 0 };
+
+    context.session = session;
+    context.callbacks = policy_ctx->exec_callbacks;
+
+    do {
+        /* We wait for file I/O to be ready if the FAPI state automata
+           are in a file I/O state. */
+        if (io.stream) {
+            r = CALL_FAPI(ifapi_io_poll(&io));
+            return_if_error(r, "Something went wrong with IO polling");
+        }
+
+        r = CALL_FAPI(ifapi_execute_tree_ex(
+            &state,
+            &context,
+            &eval_ctx,
+            NULL, /* don't use the fapi policy store *aka pstore */
+            &io,
+            NULL, /* cause it skip loading policy path */
+            &policy_ctx->policy,
+            esys_ctx,
+            policy_ctx->hash_alg,
+            false));
```

Comment:
Similar to the other NULL parameters a comment  /* Session will not be flushed in error cases */ could be added.